# 0.1 Overview

## You have a protein sequence. 

### How do you find patterns in amino acid sequence?

### How do you find the largest region of overlap between your protein and another protein?



After this course, you will be able to :
1. Obtain data from online databases programmatically
    - Specifically, PDB files from Protein Data Bank.
2. Parse the data which is in the form of a text file
    - Obtain the amino-acid sequences from the PDB file.
3. Look for patterns in the amino acid sequence
    - Plot the amino acid frequency of the protein at odd/even positions.
4. Align two protein sequences.
    - Find the largest region of overlap between 2 proteins.

# 0.2 Specific programming related objectives

1. What kind of language is `python`?
1. Different ways of running `python`
1. What is a `python` package? How to download?
1. Basic data-types of Python (`int`, `float`, `str`, `list`, `dict`)
1. Basic control elements (`for` loop, `while`, `if` statements)
1. Basic plots (histogram, line plot, colors, axis labels, legend)
1. How to do `MATLAB` or `R` stuff in Python
1. When to not use `python`

# 1. Why Python?

<img src="imgs\python_logo.png" alt="Python symbol" style="width: 200px;"/>

## Open source, general purpose, interpreted language

## Easy to read, write 
   - Fast prototyping

## Faster than MATLAB, R
   - Great with strings
   - Also do matrices, solving differential equations, interact with various databases, machine learning
   - General purpose - Build websites, GUIs, games, etc.

## There is a package to solve your problem!
   - Packages:Python::Plugins:Firefox/Chrome
   - `pip install <packagename>` e.g. `pip install numpy`


# 2. How to start with Python?

## Download a distribution (e.g. Anaconda)
- Anaconda = Python + common packages

## [Optional] Create a virtual environment 

- Avoid confusion with Python that Linux/macOS need to run
- Work on different Python versions
- Keep different projects separate

## Start coding!

# 3.1 Modes of Python

## Code files

<img src="gifs/using_code.gif" alt="Python symbol" style="width: 400px;"/>



- Long and involved projects
- Submitting jobs on cluster

# 3.2

## REPL

<img src="gifs/using_code.gif" alt="Python symbol" style="width: 400px;"/>

# TODO replace GIF

- Read-Eval-Print Loop
- Very fast prototyping, especially on cluster etc

# 3.3

## (Jupyter) Notebooks

<img src="gifs/using_code.gif" alt="Python symbol" style="width: 400px;"/>

# TODO replace GIF

- REPL + Markdown
- Share with others who want to see your code, write reports
- Reproducible research

# 3.4

## Choosing a mode

| Mode                |                                                When to use | When not to use             |
|---------------------|:-----------------------------------------------------------|-----------------------------|
| Code files          | - Long and involved codes<br> - Submitting jobs on cluster | Fast prototyping            |
| REPL                | Very fast prototyping                                      | Any serious work            |
| (Jupyter) notebooks | - Fast prototyping<br> - Embed links, images               | Long and involved codes<br> |

# 4. Jupyter notebooks

## 4.1 Uses Markdown for code + explanation in one document


![Markdown syntax](imgs\markdown_syntax.png)


## 4.2 Markdown rendered

![Markdown render](imgs\markdown_render.png)

## 4.3 Convert to other format easily
<center><img src="imgs\output_formats.png" alt="Output format" style="height: 400px;"/></center>
- Some formats may need LaTeX installed


## 4.4 Working with Jupyter

### Start a Jupyter notebook

From command line, do

```sh
jupyter notebook
```

### Jupyter notebook and virtual environments

Create environment

```sh
conda create -n myenv python=3.6
conda activate myenv
pip install package_name
```

Create a Jupyter kernel associated with environment

```sh
conda activate myenv
pip install --user ipykernel
python -m ipykernel install --user --name=myenv_kernel
```


## 4.5 Jupyter and version control

### Option 1: Notebook format, `.ipynb`
- `json` format
- More verbose

### Option 2: Convert to Rmarkdown format, `.Rmd`
- Less verbose
- Must convert back and forth

First install `jupytext package`

```sh
pip install jupytext
```

Conversions to and from `.Rmd`

```sh
jupytext --to ipynb notebook.Rmd   # convert Rmd to ipynb
jupytext --to rmarkdown notebook.ipynb   # convert ipynb to Rmd

```

# 5. Basic data types and syntax

## 5.1 Integers (`int`) & Floats (`float`)
Basically, numbers

In [1]:
example_int = 2
print(example_int)

2


In [2]:
example_float = 2.0
print(example_float)

2.0


In [3]:
print(type(example_int), type(example_float))

<class 'int'> <class 'float'>


In [4]:
x = example_int + example_float
print(x, type(x))

4.0 <class 'float'>


## 5.1.2 Basic math


In [17]:
print(1 + 2)
print(example_int - 1)

3
1


In [18]:
print(3 * 2)

6


In [19]:
print("Dividing by int is same as dividing by float", 3 / 2, 3 / 2.0)

1.5 1.5


In [22]:
print("Get quotient with // and remainder with percent sign")
print(7//4, 7%4)

Get quotient with // and remainder with percent sign
1 3


In [25]:
print("Raise to power with **")
print(2**5)

Raise to power with **
32


## 5.1.3 Mathematical functions

First import the `math` library, and call functions with `math.exp(2)` or `math.sin(3)`.

In [28]:
import math
print("Sin(pi) is : ", math.sin(math.pi))

Sin(pi) is :  1.2246467991473532e-16


Or get the NumPy package and call functions with

In [35]:
import numpy as np  # np is shorthand for numpy, so you don't have to type it out
print("Sin(pi) is : ", np.sin(np.pi))  # you can have any shorthand name you want, like `import numpy as nmp`

Sin(pi) is :  1.2246467991473532e-16


### Evaluate $(\pi + 2x)/(\sin(x/2) + \log_{10}(x))$ for x = 2

In [36]:
x = 2
y = (x + 2 * x) / (math.sin(x/2) + math.log10(x))
print(y)

5.251636631000412


## 5.2 String (`str`)

In [5]:
example_str = "abcdefghijk"
print(example_str)

abcdefghijk


In [6]:
print(type(example_str))

<class 'str'>


In [32]:
print(example_str[1], type(example_str[0]))

b <class 'str'>


In [8]:
print(example_str[0:5])

abcde


In [9]:
print(example_str[2.0])

TypeError: string indices must be integers

In [10]:
print(example_str.capitalize() + " string capitalization and addition")

Abcdefghijk string capitalization and addition


## 5.3 List (`list`)

For collections of things

In [16]:
example_list = [1, 2.0, "c"]
print(example_list, type(example_list))
print("Length of list is : ", len(example_list))

[1, 2.0, 'c'] <class 'list'>
Length of list is :  3


In [12]:
print(example_list[0:1], "but", example_list[0])

[1] but 1


In [13]:
example_list.append(421323)
print(example_list)

[1, 2.0, 'c', 421323]


In [14]:
example_list2 = example_list + ["v"]
print(f"Example list is : {example_list}")
print(f"Example list 2 is : {example_list2}")

Example list is : [1, 2.0, 'c', 421323]
Example list 2 is : [1, 2.0, 'c', 421323, 'v']


# 5.4 Boolean (`bool`)

Either `True` or `False`

# TODO, optional

# 5.x Dictionary (`dict`)

- Key-value pairs

# TODO

# Flow control statements

# Basic plotting

# `MATLAB` stuff in `python`

# TODO use Numpy/Scipy to solve a differential equation and plot it

# `R` stuff in `python`

- `pandas` Dataframe in place of R dataframe

# 6. Python 2 vs 3

- 3.x is current
- Support 2.x will stop by 2020
- Some older packages may only work with 2.x

## Key differences
| Py2                |                                                Py3 | 
|---------------------|:-----------------------------------------------------------|
| `print "this is a string"`          | `print("this is a string")` |
| `3/2` evaluates to `1`, of type `int`                | `3/2` evaluates to `1.5`, of type `float`|

# 7. Why not python?

## Slower than C, C++, Julia
   - `numba` - magically compiles Python -> C (like `MATLAB Coder`)
   - `Cython` - intermediate language with more fine-grained control
   - C extensions - actual C code used by Python (like `Rcpp` in `R` and `mex` in `MATLAB`)

## Package and python version conflicts
   - Virtual environments
   - `pipenv`

## Fewer statistical packages
  - Use R for linear models, uncommon statistical tests
  - Can call R from Python


# 8. Assignment

## A1.1 Obtain PDB File

Fill up the following function skeleton:
```python
def get_pdb_file(): # insert arguments
    print() # make print output match
    # generalize the following variables
    file_url = "https://files.rcsb.org/download/zzzz.pdb"
    download_file_name = "zzzz.pdb"
    urllib.request.urlretrieve(file_url, download_file_name) # don't modify
    print() # make print output match
    with open(download_file_name) as file_handle: # don't modify
        data = file_handle.read() # don't modify
        lines = ___ # get a list of lines
    print() # make print output match, print first line of file
```

so that when you call it:

```python
get_pdb_file("101M")
```

you get the following output:

```
Going to download : 101M
Downloaded 101M.pdb
The first line is : HEADER    OXYGEN TRANSPORT                        13-DEC-97   101M
```

## A2.1 Process file and plot amino acid frequencies

The final plot should look like this:

![Ooh a secret!](imgs\f1.png "Figure 1")


## A2.2 (challenge) Use single letter amino acid labels on X axis

The final plot should look like this:

![alt text](imgs\f2.png "Figure 2")

### Hints

1. Use the dictionary below:

```python
aa_dict = {"ala":"A", "arg":"R", "asn":"N", "asp":"D", 
           "asx":"B", "cys":"C", "glu":"E", 
           "gln":"Q","glx":"Z","gly":"G", "his":"H", 
           "ile":"I","leu":"L", "lys":"K", "met":"M",
           "phe":"F", "pro":"P", "ser":"S", "thr":"T",
           "trp":"W", "tyr":"Y", "val":"V"}
```

## A2.2 (challenge) Overlay two histograms + legend

Overlay two histograms, one of the amino acids in the even positions and another of amino acids in the odd positions. For example, if the amino acids are : 

```python
aa = ["Gly", "Ala", "Val", "Leu", "Ile", "Cys"]
```

Then make histogram of the following two lists:
```python
aa_odd = ["Gly", "Val", "Ile"]
aa_even = ["Ala", "Leu", "Cys"]
```

The final plot should look like this:

![alt text](imgs\f3.png "Figure 3")

### Hints
1. Use the `alpha` keyword argument to `plt.hist` to set transparency (0-1, 1 is opaque).


# A3 Longest common substring

1. Download the proteins `3SJU` and `IX7H`.
2. Use Shiji's function to find the longest common subsequence.

Hint
- You need to convert the 3 letter sequence to the 1 letter subsequence to use the function.

# TODO CONVERT 3 letter to 1 letter

# TODO explain longest common subsequence